Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py

```diff
@@ -90,20 +91,64 @@ def editable(cls, branch, repository=None):
         return True
 
     @classmethod
-    def branch_point(cls, repository, limit=None):
-        cnt = 0
-        commit = repository.commit(include_log=False, include_identifier=False)
-        while cls.editable(commit.branch, repository=repository):
-            cnt += 1
-            if limit and cnt > limit:
-                return None
-            commit = repository.find(argument='HEAD~{}'.format(cnt), include_log=False, include_identifier=False)
-            if cnt > 1 or commit.branch != repository.branch or cls.editable(commit.branch, repository=repository):
-                log.info('    Found {}...'.format(string_utils.pluralize(cnt, 'commit')))
-            else:
-                log.info('    No commits on editable branch')
+    def branch_point(cls, repository):
+        branches = repository.branches_for(remote=None)
+        production_branches = [
+            'remotes/{}/{}'.format(remote, branch)
+            for remote in repository.source_remotes()
+            for branch in branches[remote] if not repository.dev_branches.match(branch)
+        ]
+
+        head = run(
+            [repository.executable(), 'rev-parse', 'HEAD'],
+            cwd=repository.root_path,
+            capture_output=True,
+            encoding='utf-8',
+        ).stdout.strip()
 
-        return commit
+        partial_bases = set()
+        for shard in [
+            production_branches[cls.MERGE_BASE_SHARD_SIZE * i:cls.MERGE_BASE_SHARD_SIZE * (i + 1)]
+            for i in range(1 + len(production_branches) // cls.MERGE_BASE_SHARD_SIZE)
+        ]:
+            if not shard:
+                continue
+            result = run(
+                [repository.executable(), 'merge-base', head] + shard,
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            )
```

Comment:
I wonder if you could use xargs? Git for Windows [appears to have it](https://github.com/git-for-windows/busybox-w32/blob/main/findutils/xargs.c), so while I don't know how to invoke it from Windows, I'd think we have it available on all ports.

If so, no need for sharding logic, just run

    xargs -0 git merge-base {head}

and pipe in `'\0'.join(production_branches)`.

Reply:
I don't think `xargs` is going to help us much here. That command would end up running `git merge-base` on every ref in our list. We would still have to merge the results in a single `git merge-base` command, and we'd still have to be prepared to shard that command because it's possible that even after de-duping, we would have more than 500 merge bases. There is also the issue that running `git merge-base` 1000 times with 1000 arguments is going to be slower than running it 3 times, even if 2 of those times have 500 arguments.

Reply:
> I don't think `xargs` is going to help us much here. That command would end up running `git merge-base` on every ref in our list

I don't think so – xargs runs as few invocations as possible, respecting ARG_MAX. From xargs(1):

	     The xargs utility reads space, tab, newline and end-of-file delimited
	     strings from the standard input and executes utility with the strings as
	     arguments.
	
	     Any arguments specified on the command line are given to utility upon
	     each invocation, followed by some number of the arguments read from the
	     standard input of xargs.  This is repeated until standard input is
	     exhausted.

iiuc it does exactly what we're doing, except by counting bytes and comparing to the platform's ARG_MAX, not sharding over a fixed number of arguments. So it would potentially reduce the number of merge-base invocations needed.